controller/src/main/java/org/jboss/as/controller/ModuleIdentifierUtil.java

```diff
@@ -21,32 +23,88 @@ public final class ModuleIdentifierUtil {
      * @return the canonical representation. Will not return @{code null}
      */
     public static String canonicalModuleIdentifier(String moduleSpec) {
+        return parseModuleIdentifier(moduleSpec, ModuleIdentifierUtil::canonicalModuleIdentifier);
+    }
+
+    /**
+     * Parses the given module identifier into name and optional slot elements, passing those to the given
+     * function and returning the result of that function.
+     * <p/>
+     * This variant does not {@link #canonicalModuleIdentifier(String) canonicalize} the given identifier.
+     *
+     * @param moduleIdentifier an  identifier for a module. Cannot be {@code null}
+     * @param function a function to apply to the module's name and optional slot. Cannot be {@code null}.
+     *                 The slot value passed the function may be null if the identifier does not contain one.
+     * @return the value returned by {@code function}
+     * @param <R> the type returned by {@code function}
+     */
+    public static <R> R parseModuleIdentifier(String moduleIdentifier, BiFunction<String, String, R> function) {
+        return parseModuleIdentifier(moduleIdentifier, function, false, null);
+    }
+
+
+    /**
+     * Parses the given module identifier into name and optional slot elements, passing those to the given
+     * function and returning the result of that function.
+     * <p/>
+     *
+     * @param moduleIdentifier an identifier for a module. Cannot be {@code null}
+     * @param function a function to apply to the module's name and optional slot. Cannot be {@code null}.
+     *                 The slot value passed the function may be null if the identifier does not contain one.
```

Comment:
"passed the function"  should be "passed to the function"

Reply:
Thanks; fixed.